sbt_common/messaging/src/test/scala/uk/ac/wellcome/messaging/test/fixtures/Messaging.scala

```diff
@@ -115,22 +124,83 @@ trait Messaging
-    withLocalS3Bucket[R] and
-      withMessageReader[R] _
+  def withMessageWriter[R](bucket: Bucket,
+                           topic: Topic,
+                           writerSnsClient: AmazonSNS = snsClient)(
+    testWith: TestWith[MessageWriter[ExampleObject], R]) = {
+    val s3Config = S3Config(bucketName = bucket.name)
+    val snsConfig = SNSConfig(topicArn = topic.arn)
+    val messageConfig = MessageConfig(
+      s3Config = s3Config,
+      snsConfig = snsConfig
+    )
+
+    val messageWriter = new MessageWriter[ExampleObject](
+      messageConfig = messageConfig,
+      snsClient = writerSnsClient,
+      s3Client = s3Client,
+      keyPrefixGenerator = keyPrefixGenerator
+    )
+
+    testWith(messageWriter)
+  }
 
-  def withMessageWorkerFixtures[R] =
-    withActorSystem[R] and
-      withMetricsSender[R] _ and
-      withLocalStackSqsQueue[R] and
-      withLocalS3Bucket[R] and
-      withMessageWorker[R] _
+  def withMessageReaderFixtures[R](
+    testWith: TestWith[(Bucket, MessageReader[ExampleObject]), R]) = {
+    withLocalS3Bucket { bucket =>
+      withLocalSnsTopic { topic =>
+        withMessageReader(bucket = bucket, topic = topic) { reader =>
+          testWith((bucket, reader))
+        }
+      }
+    }
+  }
 
-  def withMessageWorkerFixturesAndMockedMetrics[R] =
-    withActorSystem[R] and
-      withMockMetricSender[R] and
-      withLocalStackSqsQueue[R] and
-      withLocalS3Bucket[R] and
-      withMessageWorker[R] _
+  def withMessageWorkerFixtures[R](
+    testWith: TestWith[(metrics.MetricsSender,
+                        Queue,
+                        Bucket,
+                        ExampleMessageWorker),
+                       R]) = {
+    withActorSystem { actorSystem =>
+      withMetricsSender(actorSystem) { metricsSender =>
+        withLocalStackSqsQueue { queue =>
+          withMessageReaderFixtures {
+            case (bucket, messageReader) =>
+              withMessageWorker(
+                actorSystem,
+                metricsSender,
+                queue,
+                messageReader) { worker =>
+                testWith((metricsSender, queue, bucket, worker))
+              }
+          }
+        }
+      }
+    }
+  }
 
+  def withMessageWorkerFixturesAndMockedMetrics[R](
+    testWith: TestWith[(metrics.MetricsSender,
+                        Queue,
+                        Bucket,
+                        ExampleMessageWorker),
+                       R]) = {
+    withActorSystem { actorSystem =>
+      withMockMetricSender { metricsSender =>
+        withLocalStackSqsQueue { queue =>
+          withMessageReaderFixtures {
+            case (bucket, messageReader) =>
+              withMessageWorker(
+                actorSystem,
+                metricsSender,
+                queue,
+                messageReader) { worker =>
+                testWith((metricsSender, queue, bucket, worker))
+              }
+          }
+        }
+      }
+    }
+  }
```

Comment:
This takes up more lines that it used to, but I broke the old code and I couldn’t work out how to fix it without just guessing.

This version is more verbose, but substantially easier to read (for me, at least!).

Reply:
And _understand_ for all of us I think.